### ISSUE TITLE
Add IDisposable adapter for any type for using statement.

### ISSUE THREAD

**Issue by bradphelan** [dotnet/roslyn#10361]
Currently it is painful to adapt classes to the using statement. If I have a class

```
public class MyFile {
    public string Path { get; };
    public void Close();
}
```

I can't use it in a using statement because it isn't IDisposable but the Close method is probably the one I am interested in. Normally I would do

```
var file = new MyFile("c:\temp\foo.txt");
using(Disposable.Create(()=>file.Close()){
    // Do stuff with file
}
```

There are two problems above. __file__ is scoped too wide. I would have to add another layer of noisy brackets. The other is the noisy __Disposable.Create__. My proposal is that for objects that do not implement IDisposable we add an overload for using.

```
using(var file = new MyFile("c:\temp\foo.txt"), ()=> file.Close()){
    // Do stuff with file
}
```
or just pass the method group

```
using(var file = new MyFile("c:\temp\foo.txt"), file.Close){
    // Do stuff with file
}
```

**Comment by svick** [comment]
This doesn't look very useful to me, since types like your MyFile should be very rare. Or do you have some specific situation where they are common?

**Comment by GeirGrusom** [comment]
You save no characters typing this out, so it seems like a performance optimization?

Also related issue #8115

**Comment by DavidArno** [comment]
Like @svick, I'm sceptical of the real need for this. How often do you really encounter such classes, without being able to access the source and thus fix them by making them implement `IDisposable`?

Having said that, if there is a real-world case for having to do what you describe, then your poroposal seems a neat way of dealing with it by both controlling the scope of `var` and by making the syntax less noisy.

**Comment by matthid** [comment]
There are already some language features that help:

```csharp
public class MyFile
{
    public string Path => "Path";
    public void Close() => Console.WriteLine("closed");
}
public class Disposable<T> : IDisposable
{
    public T Item { get; }
    private readonly Action<T> cleanup;
    public Disposable(T item, Action<T> cleanup)
    {
        this.Item = item;
        this.cleanup = cleanup;
    }
    public void Dispose()
     => this.cleanup(this.Item);
}
public static class Disposable
{
    public static Disposable<T> Create<T>(T item, Action<T> cleanup)
      => new Disposable<T>(item, cleanup);
}
using (var d = Disposable.Create(new MyFile(), f => f.Close())) {
    Console.WriteLine($"Path: {d.Item.Path}");
}

using static Disposable;
using (var d = Create(new MyFile(), f => f.Close())) {
    Console.WriteLine($"Path: {d.Item.Path}");
}
```

IMHO there is no need to bake that into the language. Of course you can come up with a better name than `Create` but both of your problems seem to be solved with this approach?

**Comment by GSPP** [comment]
Maybe `using` could make use of extension methods called `Dispose`.

Otherwise I'm voting down because this is super rare in my experience. The existence of a `Close` method is often an anti-pattern. Almost all the `Close` methods in the BCL are API design errors from the .NET 1.0 days. There's a story somewhere on the web on why they even exist.

**Comment by bradphelan** [comment]
This gives me an idea. The 'using' statement is too rigidly coupled to the
'IDisposable' interface. Let's create an interface that describes exactly
what using does. It introduces a variable into scope and it disposes
something. No reason that the thing disposed and the thing introduced into
scope are the same object. So ...

  interface IUsing<T> : IDisposable {
       T Item { get; }
  }

And a helper class

public class Using<T> {
   public T Item {get;}
   private readonly IDisposable Cleanup;
   public Using(T item, IDisposable Cleanup){
      Item = item;
      this.Cleanup = cleanup;
   }
   public void Dispose() =>  Cleanup();
}

  public static class Using {
      public static IUsing MakeUsable<T>( T item, IDisposable Cleanup) =>
new Using<T>(item,cleanup);
      public static IUsing MakeUsable<T> ( T item, Action Cleanup) =>
  MakeUsable(item, Disposable.Create(Cleanup);
     public static IUsing MakeUsable<T> (T item, Action<T> Cleanup) =>
  MakeUsable(item, () => Cleanup(item));

  }

and a simple if naive use case

  using static Using;

  using (var file = MakeUsable(new MyFile(), f => f.Close()){
      Console.WriteLine($"Path: {f.Path}");
  }


it would be more useful to be done like


  IUsing<MyFile> OpenFile(string path) =>
    MakeUsable(new MyFile(), f=>f.Close());


and

  using (var myFile = OpenFile("c:\passwords")){
    Console.WriteLine($"Path: {f.Path}");
  }

And for those that say this is an anti pattern. Often with windowing
toolkits you can show a windows. You can hide it or close it. Ie there are
two methods, Hide() and Close(). Which one of those should be the
disposable one. Depends on what you are trying to do.




Now this is very similar to what you suggested but the *using* statement
would be made aware of the interface *IItemInterface *

**Comment by GSPP** [comment]
This windowing Close method is not about disposal. It is appropriately named. Close in order to dispose of resources as a synonym for Dispose is the anti-pattern.

**Comment by bradphelan** [comment]
What is a "resource" ?

**Comment by GSPP** [comment]
Something that is a significant cost to the system while it is not disposed. When disposed it permanently transitions to a destroyed state and stops being a cost. I think this is what everyone intuitively understands to be a resource.

A window that can be reopened is not a resource. It's a two state system that can be transitioned arbitrarily.

Destroying a window is a permanent relief of resources. That's disposal.

**Comment by bradphelan** [comment]
No. A "resource" at least as far as the "using" concept is concerned is just a simple state machine for a common pattern. The pattern is that the code performs action "A", enters a scope and __must__ execute action "B" when it exits the scope to maintain correctness when it exits.  When it enters the scope some data is provided that is only valid within that scope. Within that definition you are free to do what you want. There is no requirement that anything must occupy "significant" resources. What doesn't significant mean anyway?

**Comment by GeirGrusom** [comment]
> I don't think TransactionScope fits your definition of resource

TransactionScope's resource is the transaction it's working with which must be committed when the scope exits.

**Comment by bradphelan** [comment]
My point exactly. It has nothing to do with system resources or memory or
stuff that takes up "significant" something or other. Deciding to use
"Using" is about maintaining program correctness not about deciding if some
concept of resource meets an abitrary level of significant or not.

**Comment by DavidArno** [comment]
@GSPP,

> This windowing Close method is not about disposal. It is appropriately named. Close in order to dispose of resources as a synonym for Dispose is the anti-pattern.
This is circular reasoning. `using` requires a `IDisposable`, not a `IPostUsingActionable`, therefore if one wishes to eg use `using` to simplify the code around opening and closing a window, one it is only an anti-pattern because of the design of `using`. So if there's an anti-pattern in there, it's `using` itself.

@bradphelan,
I'm still unconvinced of the need to change the behaviour of `using` though. Your `Using<T>` solution solves it without a compiler change. Why not package that code up as a NuGet package for all to use and accept it as "job done"?

**Comment by GSPP** [comment]
I think I misunderstood a bit what was being talked about, and I was misunderstood as well.

I think the issues brought up in this ticket are totally valid but so far no solution is known that should actually make it into the language, IMHO.

Can we add a `Disposable.Create` helper to the framework? I have always been missing that and needed to write it myself. I think that gets us 50% of the benefits already at minimal cost.

**Comment by HaloFour** [comment]
@GSPP 

Rx has some great `Disposable` helpers, including `Disposable.Create()`.

@bradphelan 

I proposed #11420 which would allow for the C# compiler to consider instance methods if the target type does not implement `IDisposable`.  This would allow for you to write an extension method called `Dispose` for that type which would adapt the type to the disposable pattern:

```cs
public class MyFile {
    public string Path { get; };
    public void Close();
}

public static class MyFileExtensions {
    public static void Dispose(this MyFile file) {
        file.Close();
    }
}

using(var file = new MyFile("c:\temp\foo.txt")){
    // Do stuff with file
}
```
